src/Storages/MergeTree/PartMovesBetweenShardsOrchestrator.cpp

```diff
@@ -0,0 +1,408 @@
+
+    try
+    {
+        entry_node_holder = zkutil::EphemeralNodeHolder::create(entry_to_process->znode_path + "/lock_holder", *zk, storage.replica_name);
+    }
+    catch (const Coordination::Exception & e)
+    {
+        if (e.code == Coordination::Error::ZNODEEXISTS)
+        {
+            LOG_DEBUG(log, "Task {} is being processed by another replica", entry_to_process->znode_name);
+            return false;
+        }
+
+        throw;
+    }
+
+    try
+    {
+        /// Use the same ZooKeeper connection. If we'd lost the lock then connection
+        /// will become expired and all consequent operations will fail.
+        stepEntry(entry_to_process.value(), zk);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(log, __PRETTY_FUNCTION__);
+
+        Entry entry_copy = entry_to_process.value();
+        entry_copy.last_exception_msg = getCurrentExceptionMessage(false);
+        entry_copy.update_time = std::time(nullptr);
+        zk->set(entry_copy.znode_path, entry_copy.toString(), entry_copy.version);
+
+        return false;
+    }
+
+    return true;
+}
+
+void PartMovesBetweenShardsOrchestrator::stepEntry(const Entry & entry, zkutil::ZooKeeperPtr zk)
+{
+    switch (entry.state.value)
+    {
+        case EntryState::DONE:
+            break;
+
+        case EntryState::CANCELLED:
+            break;
+
+        case EntryState::TODO:
+        {
+            /// State transition.
+            Entry entry_copy = entry;
+            entry_copy.state = EntryState::SYNC_SOURCE;
+            entry_copy.update_time = std::time(nullptr);
+            zk->set(entry_copy.znode_path, entry_copy.toString(), entry_copy.version);
+        }
+        break;
+
+        case EntryState::SYNC_SOURCE:
+        {
+            {
+                /// Log entry.
+                Coordination::Requests ops;
+                ops.emplace_back(zkutil::makeCheckRequest(entry.znode_path, entry.version));
+
+                ReplicatedMergeTreeLogEntryData log_entry;
+                log_entry.type = ReplicatedMergeTreeLogEntryData::SYNC_PINNED_PART_UUIDS;
+                log_entry.create_time = std::time(nullptr);
+                log_entry.source_replica = storage.replica_name;
+                ops.emplace_back(zkutil::makeSetRequest(zookeeper_path + "/log", "", -1));
+                ops.emplace_back(zkutil::makeCreateRequest(
+                    zookeeper_path + "/log/log-", log_entry.toString(), zkutil::CreateMode::PersistentSequential));
+
+                Coordination::Responses responses;
+                Coordination::Error rc = zk->tryMulti(ops, responses);
+                zkutil::KeeperMultiException::check(rc, ops, responses);
+
+                String log_znode_path = dynamic_cast<const Coordination::CreateResponse &>(*responses.back()).path_created;
+                log_entry.znode_name = log_znode_path.substr(log_znode_path.find_last_of('/') + 1);
+
+                storage.waitForAllReplicasToProcessLogEntry(log_entry, true);
```

Comment:
It's better to have some callback here, which will call `step` function when all replicated will be ready. This waiting in the background looks useless.